### PR TITLE
Spi code unification [CN-113]

### PR DESCRIPTION
**Pull request by SeriyBg**
Unify the code for the SPI plugins

Breaking changes:
- For the Kubernetes SPI the exception class for the failure in generating SSLSocketFactory is changed from `KubernetesClientException` to `RestClientException`

Checklist:
- [+] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [+] Add `Add to Release Notes` label if changes should be mentioned in release notes or `Not Release Notes content` if changes are not relevant for release notes
- [+] Request reviewers if possible
- [+] New public APIs have `@Nonnull/@Nullable` annotations
- [+] New public APIs have `@since` tags in Javadoc
- [+] Send backports/forwardports if fix needs to be applied to past/future releases

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/internal/util/StringUtil.java [hunk 1]
```diff
@@ -114,14 +114,58 @@ public static boolean isNullOrEmpty(String s) {
...
         }
         return s.trim().isEmpty();
     }
 
+    /**
+     * Check if all Strings are not blank
+     * @param values the strings to check
+     * @return true if all the strings are not {@code null} and not blank, false otherwise
+     */
+    public static boolean isAllNotBlank(String... values) {
```
Comment (leszko) [inline hazelcast/src/main/java/com/hazelcast/internal/util/StringUtil.java:129]: Maybe we should call it similar to other methods, like `isAllNullOrEmptyAfterTrim()`?

File: hazelcast/src/main/java/com/hazelcast/internal/util/StringUtil.java [hunk 2]
```diff
@@ -114,14 +114,58 @@ public static boolean isNullOrEmpty(String s) {
...
+        }
+        return true;
+    }
+
+    /**
+     * Check if any String from the provided Strings
+     * @param values the strings to check
+     * @return true if at least one string of the {@param values} are not {@code null} and not blank
+     */
+    public static boolean isAnyNotBlank(String... values) {
```
Comment (leszko) [inline hazelcast/src/main/java/com/hazelcast/internal/util/StringUtil.java:146]: Maybe we should call it similar to other methods, like `isAnyNullOrEmptyAfterTrim()`?

File: hazelcast/src/main/java/com/hazelcast/internal/util/StringUtil.java [hunk 3]
```diff
@@ -114,14 +114,58 @@ public static boolean isNullOrEmpty(String s) {
...
         return s.trim().isEmpty();
     }
 
+    /**
+     * Check if all Strings are not blank
+     * @param values the strings to check
+     * @return true if all the strings are not {@code null} and not blank, false otherwise
+     */
+    public static boolean isAllNotBlank(String... values) {
+        if (values == null) {
```
Comment (leszko) [inline hazelcast/src/main/java/com/hazelcast/internal/util/StringUtil.java:130]: Wouldn't this code work the same?
```
        if (values == null) {
            return false;
        }
        return !Arrays.stream(values).anyMatch(StringUtil::isNullOrEmptyAfterTrim);
```

File: hazelcast/src/main/java/com/hazelcast/internal/util/StringUtil.java [hunk 4]
```diff
@@ -114,14 +114,58 @@ public static boolean isNullOrEmpty(String s) {
...
+    }
+
+
+    /**
+     * Checks if a string is not blank or not null
+     *
+     * @param s the string to check.
+     * @return true if the string is not blank or not {@code null}, false otherwise
+     */
+    public static boolean isNotBlank(String s) {
```
Comment (leszko) [inline hazelcast/src/main/java/com/hazelcast/internal/util/StringUtil.java:165]: Do we need this method at all? Why not to use `isNullOrEmptyAfterTrim()` everywhere?

File: hazelcast/src/main/java/com/hazelcast/internal/util/StringUtil.java [hunk 5]
```diff
@@ -114,14 +114,58 @@ public static boolean isNullOrEmpty(String s) {
...
+        return true;
+    }
+
+    /**
+     * Check if any String from the provided Strings
+     * @param values the strings to check
+     * @return true if at least one string of the {@param values} are not {@code null} and not blank
+     */
+    public static boolean isAnyNotBlank(String... values) {
+        if (values == null) {
```
Comment (leszko) [inline hazelcast/src/main/java/com/hazelcast/internal/util/StringUtil.java:147]: Wouldn't the following code work the same?
```
        if (values == null) {
            return false;
        }
        return Arrays.stream(values).anyMatch(StringUtil::isNotBlank);
```

File: hazelcast/src/main/java/com/hazelcast/kubernetes/KubernetesRestClient.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateFactory;
+import java.util.Collection;
+
+/**
+ * Utility class for making REST calls.
+ */
+final class KubernetesRestClient extends AbstractRestClient<KubernetesRestClient> {
+
+    private static final ILogger LOGGER = Logger.getLogger(KubernetesRestClient.class);
```
Comment (leszko) [inline hazelcast/src/main/java/com/hazelcast/kubernetes/KubernetesRestClient.java:45]: This constant is never used.

File: hazelcast/src/main/java/com/hazelcast/kubernetes/KubernetesRestClient.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+import java.security.KeyStore;
+import java.security.cert.Certificate;
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateFactory;
+import java.util.Collection;
+
+/**
+ * Utility class for making REST calls.
+ */
+final class KubernetesRestClient extends AbstractRestClient<KubernetesRestClient> {
```
Comment (leszko) [inline hazelcast/src/main/java/com/hazelcast/kubernetes/KubernetesRestClient.java:43]: I wonder if we need to have this class at all? I think all the code is generic, in a sense that could be used as well outside of the Kubernetes context. What if we put everything in the `RestClient` class and in the `buildHttpConnection` we check add the following check? Wouldn't it work?
```
        if (connection instanceof HttpsURLConnection && caCertificate != null) {
            ((HttpsURLConnection) connection).setSSLSocketFactory(buildSslSocketFactory());
        }
```

The thing is that from the class structure someone could think that Kubernetes requires some special Rest client interaction, while in fact Kubernetes is just a standard Rest Client, it just needs mutual TLS (which is not used in at the moment in AWS/GCP/Azure).

**Review by leszko — COMMENTED** [review]
Thanks for the PR @SeriyBg . Added a few comments.

File: hazelcast/src/main/java/com/hazelcast/kubernetes/KubernetesRestClient.java (hunk 7 above)
Comment (SeriyBg) [reply hazelcast/src/main/java/com/hazelcast/kubernetes/KubernetesRestClient.java:43]: Yes, but the `buildSslSocketFactory` throws `KubernetesClientException`, and changing it to throw `RestClientExcetion` will break backward compatibility. If we are ok with this change, then we can move to one class and unify it

File: .github/workflows/aws-terraform-integration-tests.yml [hunk 8]
```diff
@@ -5,9 +5,9 @@ on:
     branches:
       - main
     paths:
-      - 'hazelcast/src/main/java/com/hazelcast/aws/**'
+      - 'hazelcast/src/main/java/com/hazelcast/gcp/**'
```
[the remark this answers is not in the document]
Comment (SeriyBg) [reply .github/workflows/aws-terraform-integration-tests.yml:8]: Oops, yeap, my bad

**Comment by SeriyBg** [comment]
@Serdaro please not one breaking change in the PR description

**Review by leszko — APPROVED** [review]
Thanks again for the PR and all the changes @SeriyBg LGTM 👍

**Pull request merged**